Cannot save classes that inherit from another with ids generic types

### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [ ] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Create an abstract class with the primary key as Generic Type extends Serializable
2. Create another class that extends the above abstract class
3. Try to save it

### Expected Behaviour

I expect that the concrete class will save in the DB (in my case is a MySql DB). With Micronaut Data 1.0.0 works fine

### Actual Behaviour

When I try to save I have the following exception.

```
>io.micronaut.data.exceptions.DataAccessException: SQL Error executing INSERT: Conversion not supported for type java.io.Serializable
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.lambda$persistOne$9(DefaultJdbcRepositoryOperations.java:661)
	at io.micronaut.transaction.support.AbstractSynchronousTransactionManager.executeWrite(AbstractSynchronousTransactionManager.java:177)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.persistOne(DefaultJdbcRepositoryOperations.java:612)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.persist(DefaultJdbcRepositoryOperations.java:602)
	at io.micronaut.data.runtime.intercept.DefaultSaveEntityInterceptor.intercept(DefaultSaveEntityInterceptor.java:43)
	at io.micronaut.data.intercept.DataIntroductionAdvice.intercept(DataIntroductionAdvice.java:80)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
	at io.micronaut.validation.ValidatingInterceptor.intercept(ValidatingInterceptor.java:123)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
	at com.navent.finprop.crm.repositories.PetitionerCRMRepositorySpec.Test save methods(PetitionerCRMRepositorySpec.groovy:19)
Caused by: java.sql.SQLException: Conversion not supported for type java.io.Serializable
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:545)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:513)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:505)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:479)
	at com.mysql.cj.jdbc.result.ResultSetImpl.getObject(ResultSetImpl.java:1382)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.getEntityId(DefaultJdbcRepositoryOperations.java:1118)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.lambda$persistOne$9(DefaultJdbcRepositoryOperations.java:639)
	... 9 more
```

I have the following structures.

**The abstract base class**

```
public abstract class BaseEntity<ID extends Serializable> implements Serializable {

    @Id
    @GeneratedValue
    private ID id;
    @DateCreated
    private LocalDateTime createdDate;
    @DateUpdated
    private LocalDateTime updatedDate;
}
```

**The concrete class**

```
@Entity
public class Guarantee extends BaseEntity<Long> {

    @NotBlank
    @Size(max = 45)
    private String code;
    @NotNull
    private GuaranteeStatus status;
    private BigDecimal price;
}

```

### Environment Information

- **Operating System**: Ubuntu 18.04 LTS
- **Micronaut Version:** mn 1.3.3 and mn-data 1.0.2
- **JDK Version:** 
openjdk version "11.0.6" 2020-01-14
OpenJDK Runtime Environment 18.9 (build 11.0.6+10)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.6+10, mixed mode)


### Example Application

- TODO: link to github repository with example that reproduces the issue

--- a/inject-java-test/src/test/groovy/io/micronaut/inject/visitor/beans/BeanIntrospectionSpec.groovy
+++ b/inject-java-test/src/test/groovy/io/micronaut/inject/visitor/beans/BeanIntrospectionSpec.groovy
@@ -87,6 +87,38 @@ interface GenBase<T> {
         def test = introspection.instantiate()
 
         then:
+        introspection.beanProperties.first().type == String
+        introspection.getRequiredProperty("name", String)
+                .get(test) == 'test'
+    }
+
+    void "test bean introspection with property of generic superclass"() {
+        given:
+        BeanIntrospection introspection = buildBeanIntrospection('test.Foo', '''
+package test;
+
+@io.micronaut.core.annotation.Introspected
+class Foo extends GenBase<String> {
+    public String getName() {
+        return "test";
+    }
+}
+
+abstract class GenBase<T> {
+    abstract T getName();
+    
+    public T getOther() {
+        return (T) "other";
+    }
+}
+''')
+        when:
+        def test = introspection.instantiate()
+
+        def beanProperties = introspection.beanProperties.toList()
+        then:
+        beanProperties[0].type == String
+        beanProperties[1].type == String
         introspection.getRequiredProperty("name", String)
                 .get(test) == 'test'
     }

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/visitor/JavaClassElement.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/visitor/JavaClassElement.java
@@ -296,6 +296,14 @@ public class JavaClassElement extends AbstractJavaElement implements ClassElemen
                             value.type,
                             value.setter == null,
                             visitorContext) {
+
+                        @Override
+                        public ClassElement getGenericType() {
+                            TypeMirror returnType = value.getter.getReturnType();
+                            Map<String, Map<String, TypeMirror>> declaredGenericInfo = getGenericTypeInfo();
+                            return parameterizedClassElement(returnType, visitorContext, declaredGenericInfo);
+                        }
+
                         @Override
                         public Optional<String> getDocumentation() {
                             Elements elements = visitorContext.getElements();

--- a/inject/src/main/java/io/micronaut/inject/beans/visitor/IntrospectedTypeElementVisitor.java
+++ b/inject/src/main/java/io/micronaut/inject/beans/visitor/IntrospectedTypeElementVisitor.java
@@ -31,6 +31,8 @@ import io.micronaut.inject.writer.ClassGenerationException;
 
 import edu.umd.cs.findbugs.annotations.NonNull;
 import edu.umd.cs.findbugs.annotations.Nullable;
+import org.jetbrains.annotations.NotNull;
+
 import java.io.IOException;
 import java.util.*;
 import java.util.concurrent.atomic.AtomicInteger;
@@ -351,9 +353,16 @@ public class IntrospectedTypeElementVisitor implements TypeElementVisitor<Object
         writers.put(writer.getBeanType().getClassName(), writer);
     }
 
-    private void processBeanProperties(BeanIntrospectionWriter writer, Collection<? extends PropertyElement> beanProperties, Set<String> includes, Set<String> excludes, Set<String> ignored, Set<AnnotationValue> indexedAnnotations, boolean metadata) {
+    private void processBeanProperties(
+            BeanIntrospectionWriter writer,
+            Collection<? extends PropertyElement> beanProperties,
+            Set<String> includes,
+            Set<String> excludes,
+            Set<String> ignored,
+            Set<AnnotationValue> indexedAnnotations,
+            boolean metadata) {
         for (PropertyElement beanProperty : beanProperties) {
-            final ClassElement type = beanProperty.getType();
+            final ClassElement type = beanProperty.getGenericType();
 
             final String name = beanProperty.getName();
             if (!includes.isEmpty() && !includes.contains(name)) {

